crates/runtime/src/libcalls.rs

```diff
@@ -487,6 +491,144 @@ unsafe fn new_epoch(instance: &mut Instance) -> Result<u64> {
+                let result = valgrind_state.read(addr as usize + offset as usize, num_bytes as usize);
+                match result {
+                    Ok(()) => {
+                        return Ok(0);
+                    }
+                    Err(InvalidRead { addr, len }) => {
+                        bail!("Invalid load at addr {:#x} of size {}", addr, len);
+                    }
+                    Err(OutOfBounds { addr, len }) => {
+                        bail!("Load out of bounds at addr {:#x} of size {}", addr, len);
+                    }
+                    _ => {
+                        panic!("unreachable")
+                    }
+                }
+            }
+            Ok(0)
+        }
+
+        // Hook for validating store using valgrind_state.
+        fn check_store(instance: &mut Instance, num_bytes: u32, addr: u32, offset: u32) -> Result<u32> {
+            if let Some(valgrind_state) = &mut instance.valgrind_state {
+                let result = valgrind_state.write(addr as usize + offset as usize, num_bytes as usize);
+                match result {
+                    Ok(()) => {
+                        return Ok(0);
+                    }
+                    Err(InvalidWrite { addr, len }) => {
+                        bail!("Invalid store at addr {:#x} of size {}", addr, len)
+                    }
+                    Err(OutOfBounds { addr, len }) => {
+                        bail!("Store out of bounds at addr {:#x} of size {}", addr, len)
+                    }
+                    _ => {
+                        panic!("unreachable")
+                    }
+                }
+            }
+            Ok(0)
+        }
+
+        // Hook for turning valgrind load/store validation off when entering a malloc function.
+        fn malloc_start(instance: &mut Instance) {
+            if let Some(valgrind_state) = &mut instance.valgrind_state {
+                valgrind_state.memcheck_off();
+            }
+        }
+
+        // Hook for turning valgrind load/store validation off when entering a free function.
+        fn free_start(instance: &mut Instance) {
+            if let Some(valgrind_state) = &mut instance.valgrind_state {
+                valgrind_state.memcheck_off();
+            }
+        }
+
+        // Hook for tracking wasm stack updates using valgrind_state.
+        fn update_stack_pointer(instance: &mut Instance, value: u32) {
+            if let Some(valgrind_state) = &mut instance.valgrind_state {
+                // instance.valgrind_state.update_stack_pointer(value as usize);
+            }
+        }
+
+        // Hook updating valgrind_state memory state vector every time memory.grow is called.
+        fn update_mem_size(instance: &mut Instance, num_pages: u32) {
+            if let Some(valgrind_state) = &mut instance.valgrind_state {
+                let kib: usize = 1024;
+                let num_bytes = num_pages as usize * kib * 64;
```

Comment:
enormously tiny nitpick but this would read slightly more naturally for me as `... * 64 * KIB`